README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# howzatR

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/howzatR)](https://CRAN.R-project.org/package=howzatR)
<!-- badges: end -->

The goal of howzatR is to provide useful functions for cricket analysis
& exploratory.

## Installation

You can install a stable version of howzatR using R/Rstudio with:

``` r
install.packages("howzatR")
```

You can install the development version of howzatR from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("lukelockley/howzatR")
```

## Example - Batters Analysis

This is a basic example how to use the batting functionality:

``` r
library(howzatR)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

## Basic Batting dataset
bat_raw_df
#>     Player Inns NO Runs_Scored Balls_Faced
#> 1 A. Green    7  1         140         220
#> 2 B. Brown    8  3         156         100
#> 3  C. Blue    6  0         111          76


## Analysis
bat_df <- bat_raw_df %>%
  mutate(
    Outs = Inns - NO,
    Average = bat_avg(runs_scored = Runs_Scored, no_dismissals = Outs),
    Strike_Rate = bat_sr(runs_scored = Runs_Scored, balls_faced = Balls_Faced)
  )

## Results
bat_df
#>     Player Inns NO Runs_Scored Balls_Faced Outs  Average Strike_Rate
#> 1 A. Green    7  1         140         220    6 23.33333    63.63636
#> 2 B. Brown    8  3         156         100    5 31.20000   156.00000
#> 3  C. Blue    6  0         111          76    6 18.50000   146.05263
```

## Example - Bowling Analysis

This is a basic example how to use the bowling functionality

``` r
library(howzatR)
library(dplyr)

## Basic Bowling dataset
bowl_raw_df
#>     Player Balls_Bowled Runs_Conceded Wickets
#> 1 E. Apple          560           235      15
#> 2  F. Pear          754           567      21
#> 3 G. Grape          234           270       7


## Analysis
bowl_df <- bowl_raw_df %>%
  mutate(
    Economy_overs = bowl_econ(balls_bowled = Balls_Bowled, runs_conceded = Runs_Conceded, type = "overs"),
    Economy_sets = bowl_econ(balls_bowled = Balls_Bowled, runs_conceded = Runs_Conceded, type = "sets"),
    Economy_hundred = bowl_econ(balls_bowled = Balls_Bowled, runs_conceded = Runs_Conceded, type = "per_100"),
    Average = bowl_avg(runs_conceded = Runs_Conceded, wickets_taken = Wickets),
    Strike_Rate = bowl_sr(balls_bowled = Balls_Bowled, wickets_taken = Wickets),
    Overs = balls_to_overs(balls = Balls_Bowled)
  ) %>%
  select(
    Player, Balls_Bowled, Overs, Runs_Conceded,
    Wickets, Economy_overs, Economy_sets, Economy_hundred,
    Average, Strike_Rate
  )

## Results
bowl_df
#>     Player Balls_Bowled Overs Runs_Conceded Wickets Economy_overs Economy_sets
#> 1 E. Apple          560  93.2           235      15      2.517857     2.098214
#> 2  F. Pear          754 125.4           567      21      4.511936     3.759947
#> 3 G. Grape          234  39.0           270       7      6.923077     5.769231
#>   Economy_hundred  Average Strike_Rate
#> 1        41.96429 15.66667    37.33333
#> 2        75.19894 27.00000    35.90476
#> 3       115.38462 38.57143    33.42857
```


NEWS.md
# howzatR (development version)

# howzatR 1.0.1

* Update package `DESCRIPTION` with `URL` and `BugReports` entries
* Add how to install stable version of howzatR in `README` 
* Update `bowl_econ` function to include parameter `type` to allow to calculate Economy Rate per six ball overs, five ball sets or hundred balls bowled
* Update `README` to include examples of the above

# howzatR 1.0.0

* Added a `NEWS.md` file to track changes to the package after the first release.


tests/testthat.R
library(testthat)
library(howzatR)

test_check("howzatR")


tests/testthat/test-balls_overs.R

# balls_to_overs ----------------------------------------------------------

test_that("balls_to_overs work", {
  expect_equal(balls_to_overs(balls = 10), 1.4)
  expect_equal(balls_to_overs(balls = 10 * 6), 10)
})


# overs_to_balls ----------------------------------------------------------

test_that("overs_to_balls work", {
  expect_equal(overs_to_balls(overs = 10.2), 62)
  expect_equal(overs_to_balls(overs = 10 * 6), 60 * 6)
})

test_that("overs_to_balls greater than 6 Balls supplied causes error", {
  balls_no <- 8

  expect_error(overs_to_balls(overs = 10.8),
    regexp = paste(
      "Oops... an over can only contain 6 legitimate balls. You supplied",
      balls_no, "balls which is impossible. Please check your input!"
    )
  )
})


tests/testthat/test-helper_funcs.R

# calc_int_div & calc_mod -------------------------------------------------

test_that("calc_int_div & calc_mod works", {
  expect_equal(calc_mod(value = 5.5, divisor = 1), 0.5)
  expect_equal(calc_int_div(value = 5.5, divisor = 1), 5)
})


R/balls_overs.R
# Balls to Overs ----------------------------------------------------------

#' Convert Balls to Overs (Six Ball)
#'
#' Convert numbers of balls as it equates in terms of six ball overs.
#'
#' @param balls number of balls bowled/faced.
#'
#' @return number of six ball overs this equates too.
#' @export
#'
#' @examples
#' balls_to_overs(balls = 6)
#' balls_to_overs(balls = 17)
#'
balls_to_overs <- function(balls) {

  # get over number using integer division
  over_no <- calc_int_div(value = balls, divisor = 6)

  # get balls remaining in over using modulo
  balls_no <- calc_mod(value = balls, divisor = 6)


  over_no + (balls_no / 10)
}


# Overs to Balls ----------------------------------------------------------

#' Convert Overs (Six Ball) to Balls
#'
#' @param overs number of six ball overs bowled/faced.
#'
#' @return number of six ball overs this equates too.
#' @export
#'
#' @examples
#' overs_to_balls(overs = 8.2)
#' overs_to_balls(overs = 10)
#'
overs_to_balls <- function(overs) {

  # get over number using integer division
  over_no <- calc_int_div(value = overs, divisor = 1)

  # get balls remaining in over using modulo
  balls_no <- calc_mod(value = overs, divisor = 1)

  if (balls_no > 0.6) {
    mes <- paste(
      "Oops... an over can only contain 6 legitimate balls. You supplied",
      as.integer(balls_no * 10),
      "balls which is impossible. Please check your input!"
    )

    rlang::abort(message = mes)
  }

  (over_no) * 6 + (balls_no) * 10
}


R/helper_funcs.R

# Modulo & Integer Division Calculation -----------------------------------

#' Calculates Modulo
#'
#' Calculates remainder of a division. See \url{https://en.wikipedia.org/wiki/Modulo_operation} for more details
#'
#' @param value number to divide through
#' @param divisor number to use or the division
#'
#' @return remainder of division
#' @noRd
#'
#' @examples
#' calc_mod(value = 5.5, divisor = 1)
calc_mod <- function(value, divisor) {
  value %% divisor
}

#' Calculates Integer Division
#'
#' Calculates division and discards the remainder. See \url{https://mathworld.wolfram.com/IntegerDivision.html} for more details
#'
#' @param value number to divide through
#' @param divisor number to use or the division
#'
#' @return division with remainder discarded; leaving just the integer
#' @noRd
#'
#' @examples
#' calc_int_div(value = 5.5, divisor = 1)
calc_int_div <- function(value, divisor) {
  value %/% divisor
}


R/batting_basics.R
# Batting Average ---------------------------------------------------------

#' Batters Average
#'
#' Calculates a batter's average over a number of innings.
#'
#' @param runs_scored A singular value of the runs scored by a batter.
#' @param no_dismissals A singular value of the number of times a batters has been dismissed within those innings.
#'
#' @return A singular value showing the batter's average.
#' @export
#'
#' @section Additional Information:
#' A batting average is the number of runs divided by the number of times a batters is dismissed.
#' Batters who remain \strong{not out} at the end of an innings \strong{don't} have that innings
#' count towards the number of dismissals. The higher average typically indicates a higher quality player.
#' More info [here.](https://en.wikipedia.org/wiki/Batting_average_(cricket))
#'
#' @examples
#' bat_avg(runs_scored = 568, no_dismissals = 9)
#' total_runs <- sum(c(45, 123, 56, 12, 192, 34, 78, 3, 25))
#' bat_avg(runs_scored = total_runs, no_dismissals = 9)
bat_avg <- function(runs_scored, no_dismissals) {
  runs_scored / no_dismissals
}


# Batting Strike Rate -----------------------------------------------------

#' Batters Strike Rate
#'
#' Calculates a batter's strike rate over a number of innings.
#'
#' @param runs_scored A singular value of the runs scored by a batter.
#' @param balls_faced A singular value of balls faced by a batter.
#'                    Overs can be converted into `balls_faced` using \code{\link{overs_to_balls}}
#'
#' @return A singular value showing the batter's strike rate per 100 Balls.
#'
#' @section Additional Information:
#' A batting strike rate is the average number of runs scored per 100 balls.
#' For example, a strike rate of 135 implies a batter would score 135 runs in a 100 balls.
#' A higher number indicates the batter scores at faster rate.
#' More info [here.](https://en.wikipedia.org/wiki/Strike_rate#Batting_strike_rate)
#'
#'
#' @export
#'
#' @examples
#' bat_sr(runs_scored = 568, balls_faced = 600)
#' total_runs <- sum(c(45, 123, 56, 12, 192, 34, 78, 3, 25))
#' total_balls <- sum(c(50, 120, 78, 3, 226, 36, 45, 12, 30))
#' bat_sr(
#'   runs_scored = total_runs,
#'   balls_faced = total_balls
#' )
bat_sr <- function(runs_scored, balls_faced) {
  (runs_scored / balls_faced) * 100
}


R/bowling_basics.R

# Bowling Economy Rate ------------------------------------------------

#' Bowler Economy Rate
#'
#' Calculates bowlers' economy rate over six ball overs, five ball sets or per hundred balls.
#'
#' @param balls_bowled number of balls bowled. Data in terms of six ball overs,
#'                     please convert to \code{\link{overs_to_balls}} to get it terms of `balls bowled`
#' @param runs_conceded total runs conceded by bowler across the overs, sets or per hundred balls bowled.
#' @param type whether we are calculating economy over six ball overs, sets or per hundred balls bowled.
#'             Options "overs", "sets", "per_100". Defaults to overs

#' @return Economy rate across the number of overs, sets or per hundred balls bowled.
#' @export
#'
#' @section Additional Information:
#' Bowling economy rate is average number of runs scored per over or sets bowled.
#'   * If using overs, a value of 9.5 indicates an average of 9.5 runs are scored per six ball over bowled.
#'   * If using sets, a value of 9.5 indicates an average of 9.5 runs are scored per five ball set bowled.
#'   * If using here, a value of 9.5 indicates an average of 9.5 runs are scored per hundred balls bowled.
#' This the official statistic used by [The Hundred.](https://www.thehundred.com/stats?stat=econRatePerHundred&type=men&year=2022) \cr
#'   \cr
#' The higher the number the more detrimental is for the bowler. Runs scored through byes & leg byes are \strong{excluded} from runs conceded by the bowler,
#' however wides and no-balls are \strong{included} in the bowler's figures. \cr
#' More info [here.](https://en.wikipedia.org/wiki/Economy_rate)
#'
#'
#'
#' @examples
#' bowl_econ(balls_bowled = 60, runs_conceded = 45)
#' bowl_econ(
#'   balls_bowled = overs_to_balls(overs = 7.1),
#'   runs_conceded = 26,
#'   type = "overs"
#' )
#'
#' bowl_econ(balls_bowled = 30, runs_conceded = 35, type = "sets")
#'
#' bowl_econ(balls_bowled = 22, runs_conceded = 19, type = "per_100")
#'
bowl_econ <- function(balls_bowled, runs_conceded, type = "overs") {

  # Named List of Balls within type (overs contains 6 Balls, sets contains 5 Balls & per_100 contains 100)
  type_list <- list(overs = 6, sets = 5, per_100 = 100)

  # lowercase type
  type <- tolower(type)

  # If type not in list than needs to be re-supplied
  if (!(type %in% names(type_list))) {
    rlang::abort(message = "Type can either be overs or sets. Please Amend!")
  }

  # Calculate Economy Rate
  type_list[[type]] * (runs_conceded / balls_bowled)
}


# Bowling Average ---------------------------------------------------------

#' Bowler Average
#'
#' Calculates bowlers' average number of runs per wicket taken across overs bowled.
#'
#' @param runs_conceded total runs conceded by bowler across the overs bowled.
#' @param wickets_taken total wickets taken across the overs bowled.
#'
#' @return Average number of runs per wicket taken across overs bowled.
#' @export
#'
#' @section Additional Information:
#' A bowling average is the average number of runs conceded for wicket taken. A value of 15 indicates an average of
#' 15 runs were conceded per wicket taken. The lower the value, the better the average; the reserve of \code{\link{bat_avg}}
#' More info [here.](https://en.wikipedia.org/wiki/Bowling_average)
#'
#'
#' @examples
#' bowl_avg(runs_conceded = 50, wickets_taken = 6)
#' bowl_avg(runs_conceded = 341, wickets_taken = 13)
bowl_avg <- function(runs_conceded, wickets_taken) {
  runs_conceded / wickets_taken
}


# Bowling Strike Rate -----------------------------------------------------

#' Bowler Strike Rate
#'
#' Calculates bowlers' number of balls per wicket taken across overs bowled.
#'
#' @param balls_bowled number of balls bowled. Data in terms of six ball overs.
#'                     please convert to \code{\link{overs_to_balls}} to get it terms of `balls bowled`
#' @param wickets_taken total wickets taken across the overs bowled.
#'
#' @return Number of balls per wicket taken across overs bowled.
#' @export
#'
#' @section Additional Information:
#' A bowling strike rate is defined as the number of legal balls per wicket taken.
#' For example a value of 20 indicates 20 balls bowled are scored per wicket. This the reverse of
#' \code{\link{bat_sr}} where the lower the number the better.
#' More info [here.](https://en.wikipedia.org/wiki/Strike_rate#Bowling_strike_rate)
#'
#' @examples
#' bowl_sr(balls_bowled = 3830, wickets_taken = 112)
#' bowl_sr(balls_bowled = overs_to_balls(overs = 1651.2), wickets_taken = 243)
bowl_sr <- function(balls_bowled, wickets_taken) {
  balls_bowled / wickets_taken
}
